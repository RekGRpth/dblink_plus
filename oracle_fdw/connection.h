/*-------------------------------------------------------------------------
 *
 * connection.h
 *		  Connection management for oracle_fdw
 *
 * Copyright (c) 2011, NIPPON TELEGRAPH AND TELEPHONE CORPORATION
 *
 * IDENTIFICATION
 *		  contrib/oracle_fdw/connection.h
 *
 *-------------------------------------------------------------------------
 */
#ifndef CONNECTION_H
#define CONNECTION_H

#include "foreign/foreign.h"
#include "oracle_fdw.h"

/*
 * Connection management
 */
ORAconn *GetConnection(ForeignServer *server, UserMapping *user);
void ReleaseConnection(ORAconn *conn);

#endif /* CONNECTION_H */
