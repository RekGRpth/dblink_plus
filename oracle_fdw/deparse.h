/*
 * deparse.h
 *
 * contrib/postgresql_fdw/deparse.h
 * Copyright (c) 2011, PostgreSQL Global Development Group
 * ALL RIGHTS RESERVED;
 *
 */

#ifndef DEPARSE_H
#define DEPARSE_H

/*
 * deparse.c: deparsing query-tree into SQL statement
 */
char *deparseSql(Oid foreigntableid, PlannerInfo *root, RelOptInfo *baserel);
void deparseBytea(StringInfo buf, Datum val);
void deparseInterval(StringInfo buf, Datum val);
void deparseDate(StringInfo buf, Datum val);
void deparseTimestamp(StringInfo buf, Datum val);
void deparseTimestamptz(StringInfo buf, Datum val);

#endif   /* DEPARSE_H */
