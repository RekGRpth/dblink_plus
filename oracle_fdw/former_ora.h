#ifndef FORMER_ORA_H
#define FORMER_ORA_H

typedef enum
{
	OCI_BIN,
	OCI_HEX,
	OCI_STR
} OciFetchType;

typedef struct TupleFormer   TupleFormer;
extern TupleFormer *TupleFormerCreate(int nfields);
extern void TupleFormerTerm(TupleFormer *former);
extern void TupleFormerSetType(TupleFormer *former, int col, OciFetchType type);

#endif   /* FORMER_ORA_H */
