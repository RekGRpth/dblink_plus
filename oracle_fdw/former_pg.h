#ifndef FORMER_PG_H
#define FORMER_PG_H

#include "access/htup.h"
#include "access/tupdesc.h"

extern void TupleFormerInit(TupleFormer *former, TupleDesc desc);
extern HeapTuple TupleFormerTuple(TupleFormer *former);
extern Datum TupleFormerValue(TupleFormer *former, const char *str, int size, int col);

#endif   /* FORMER_PG_H */
