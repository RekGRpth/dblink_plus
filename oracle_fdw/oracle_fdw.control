# oracle_fdw extension
comment = 'foreign-data wrapper for remote Oracle servers'
default_version = '1.0'
module_pathname = '$libdir/oracle_fdw'
relocatable = true
