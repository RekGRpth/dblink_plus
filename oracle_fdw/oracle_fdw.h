/*-------------------------------------------------------------------------
 *
 * oracle_fdw.h
 *		  foreign-data wrapper for remote Oracle servers.
 *
 * Copyright (c) 2011, NIPPON TELEGRAPH AND TELEPHONE CORPORATION
 *
 * IDENTIFICATION
 *		  contrib/oracle_fdw/oracle_fdw.h
 *
 *-------------------------------------------------------------------------
 */

#ifndef ORACLE_FDW_H
#define ORACLE_FDW_H

#include "libpq-fe.h"

#include "former_ora.h"

/* maximum of max_value_len. same as maximum of VARSIZE_4B */
#define MAX_DATA_SIZE		0x3FFFFFFF

/*
 * ORAconn encapsulates a connection to the Oracle.
 * The contents of this struct are not supposed to be known to applications.
 */
typedef struct ora_conn ORAconn;

/*
 * ORAresult encapsulates the result of a query.
 * The contents of this struct are not supposed to be known to applications.
 */
typedef struct ora_result ORAresult;

/* in option.c */
extern Datum oracle_fdw_validator(PG_FUNCTION_ARGS);
extern const char *GetFdwOption(Oid relid, const char *optname);
extern const char *GetFdwOptionList(List *defelems, const char *optname);
extern int parse_int_value(const char *value, const char *keyword);

/* make a new client connection to the backend */
extern ORAconn *ORAsetdbLogin(const char *user, const char *password, const char *dbname, const char *timezone, int fetchsize, int max_value_len);

/* close the current connection and free the PGconn data structure */
extern void ORAfinish(ORAconn *conn);

/* Accessor functions for ORAconn objects */
extern char *ORAdb(const ORAconn *conn);
extern char *ORAuser(const ORAconn *conn);
extern char *ORApass(const ORAconn *conn);
extern char *ORAtimezone(const ORAconn *conn);
extern int ORAfetchsize(const ORAconn *conn);
extern int ORAmax_value_len(const ORAconn *conn);
extern ConnStatusType ORAstatus(const ORAconn *conn);
extern char *ORAerrorMessage(const ORAconn *conn);

/* Cursor  */
extern ORAresult *ORAopen(ORAconn *conn, const char *sql, int fetchsize, int max_value_len);
extern bool ORAhasNext(ORAresult *cur);

/* set fetchsize */
void ORAsetFetchsize(ORAconn *conn, int fetchsize);
/* set max_value_len */
void ORAsetMaxValueLen(ORAconn *conn, int max_value_len);

/* Accessor functions for PGresult objects */
extern ExecStatusType ORAresultStatus(ORAresult *res);
extern int ORAnfields(const ORAresult *res);
extern char *ORAgetvalue(ORAresult *res, int field_num);
extern int ORAgetisnull(const ORAresult *res, int field_num);
extern int ORAgetamountsize(ORAresult *res, int field_num);

/* Delete a ORAresult */
extern void ORAclear(ORAresult *res);

/* get TupleFormer */
extern TupleFormer *ORAgetTupleFormer(ORAresult *res);

extern void ora_ereport(int sqlstate, const char *message, const char *detail, const char *hint, const char *context);
extern void ora_elog(int level, const char *message);

const char *get_timezone(void);

#endif   /* ORACLE_FDW_H */
