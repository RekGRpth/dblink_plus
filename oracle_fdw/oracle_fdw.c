/*-------------------------------------------------------------------------
 *
 * oracle_fdw.c
 *		  foreign-data wrapper for remote Oracle servers.
 *
 * Copyright (c) 2011, NIPPON TELEGRAPH AND TELEPHONE CORPORATION
 *
 * IDENTIFICATION
 *		  contrib/oracle_fdw/oracle_fdw.c
 *
 *-------------------------------------------------------------------------
 */
#include "postgres.h"
#include "fmgr.h"

#include "catalog/pg_foreign_server.h"
#include "catalog/pg_foreign_table.h"
#include "commands/explain.h"
#include "foreign/fdwapi.h"
#include "funcapi.h"
#include "miscadmin.h"
#include "optimizer/cost.h"
#include "utils/lsyscache.h"
#include "utils/memutils.h"
#include "utils/rel.h"

#include "oracle_fdw.h"
#include "former_pg.h"
#include "deparse.h"
#include "connection.h"

PG_MODULE_MAGIC;

/*
 * Default fetch size for cursor.  This can be overrideen by fetchsize FDW
 * option.
 */
#define DEFAULT_FETCHSIZE			100

/*
 * Cost to establish a connection.
 * XXX: should be configurable per server?
 */
#define CONNECTION_COSTS			100.0

/*
 * Cost to transfer 1 byte from remote server.
 * XXX: should be configurable per server?
 */
#define TRANSFER_COSTS_PER_BYTE		0.001

/*
 * Oracle specific portion of a foreign query request
 */
typedef struct OraFdwExecutionState
{
	FdwPlan	   *fdwplan;	/* FDW-specific planning information */
	ORAconn	   *conn;		/* connection for the scan */
	ORAresult   *res;		/* result of the scan, held until the scan ends */
} OraFdwExecutionState;

/*
 * SQL functions
 */
extern Datum oracle_fdw_handler(PG_FUNCTION_ARGS);
PG_FUNCTION_INFO_V1(oracle_fdw_handler);

/*
 * FDW callback routines
 */
static FdwPlan *oraPlanForeignScan(Oid foreigntableid,
								   PlannerInfo *root,
								   RelOptInfo *baserel);
static void oraExplainForeignScan(ForeignScanState *node, ExplainState *es);
static void oraBeginForeignScan(ForeignScanState *node, int eflags);
static TupleTableSlot *oraIterateForeignScan(ForeignScanState *node);
static void oraReScanForeignScan(ForeignScanState *node);
static void oraEndForeignScan(ForeignScanState *node);

/*
 * Helper functions
 */
static void estimate_costs(PlannerInfo *root,
						   RelOptInfo *baserel,
						   const char *sql,
						   Cost *startup_cost,
						   Cost *total_cost);
static void execute_query(ForeignScanState *node);
static void store_result(TupleTableSlot *slot, ORAresult *res);

/*
 * Foreign-data wrapper handler function: return a struct with pointers
 * to my callback routines.
 */
Datum
oracle_fdw_handler(PG_FUNCTION_ARGS)
{
	FdwRoutine *fdwroutine = makeNode(FdwRoutine);

	fdwroutine->PlanForeignScan = oraPlanForeignScan;
	fdwroutine->ExplainForeignScan = oraExplainForeignScan;
	fdwroutine->BeginForeignScan = oraBeginForeignScan;
	fdwroutine->IterateForeignScan = oraIterateForeignScan;
	fdwroutine->ReScanForeignScan = oraReScanForeignScan;
	fdwroutine->EndForeignScan = oraEndForeignScan;

	PG_RETURN_POINTER(fdwroutine);
}

/*
 * oraPlanForeignScan
 *		Create a FdwPlan for a scan on the foreign table
 */
static FdwPlan *
oraPlanForeignScan(Oid foreigntableid,
				   PlannerInfo *root,
				   RelOptInfo *baserel)
{
	FdwPlan	   *fdwplan;
	char	   *sql;

	/* Construct FdwPlan with cost estimates */
	fdwplan = makeNode(FdwPlan);
	sql = deparseSql(foreigntableid, root, baserel);
	estimate_costs(root, baserel, sql,
				   &fdwplan->startup_cost, &fdwplan->total_cost);

	/*
	 * Store plain SELECT statement in private area of FdwPlan.  This will be
	 * used for executing remote query and explaining scan.
	 */
	fdwplan->fdw_private = list_make1(makeString(sql));

	return fdwplan;
}

/*
 * oraExplainForeignScan
 *		Produce extra output for EXPLAIN
 */
static void 
oraExplainForeignScan(ForeignScanState *node, struct ExplainState *es)
{
	FdwPlan	   *fdwplan;
	char	   *sql;

	fdwplan = ((ForeignScan *) node->ss.ps.plan)->fdwplan;
	sql = strVal(list_nth(fdwplan->fdw_private, 0));
	ExplainPropertyText("Remote SQL", sql, es);
}

/*
 * oraBeginForeignScan
 *		Initiate access to a foreign Oracle table.
 */
static void
oraBeginForeignScan(ForeignScanState *node, int eflags)
{
	OraFdwExecutionState *festate;
	ORAconn		   *conn;
	Oid				relid;
	ForeignTable   *table;
	ForeignServer  *server;
	UserMapping	   *user;

	/*
	 * Do nothing in EXPLAIN (no ANALYZE) case.  node->fdw_state stays NULL.
	 */
	if (eflags & EXEC_FLAG_EXPLAIN_ONLY)
		return;

	/*
	 * Save state in node->fdw_state.
	 */
	festate = (OraFdwExecutionState *) palloc(sizeof(OraFdwExecutionState));
	festate->fdwplan = ((ForeignScan *) node->ss.ps.plan)->fdwplan;

	/*
	 * Get connection to the foreign server.  Connection manager would
	 * establish new connection if necessary.
	 */
	relid = RelationGetRelid(node->ss.ss_currentRelation);
	table = GetForeignTable(relid);
	server = GetForeignServer(table->serverid);
	user = GetUserMapping(GetOuterUserId(), server->serverid);
	conn = GetConnection(server, user);
	festate->conn = conn;

	/* Result will be filled in first Iterate call. */
	festate->res = NULL;

	/* Store FDW-specific state into ForeignScanState */
	node->fdw_state = (void *) festate;

	return;
}

/*
 * oraIterateForeignScan
 *		Retrieve next row from the result set, or clear tuple slot to indicate
 *		EOF.
 */
static TupleTableSlot *
oraIterateForeignScan(ForeignScanState *node)
{
	OraFdwExecutionState *festate;
	TupleTableSlot *slot = node->ss.ss_ScanTupleSlot;

	festate = (OraFdwExecutionState *) node->fdw_state;

	/*
	 * If this is the first call after Begin, we need to execute remote query.
	 * If the query needs cursor, we declare a cursor at first call and fetch
	 * from it in later calls.
	 */
	if (festate->res == NULL)
		execute_query(node);


	if (ORAhasNext(festate->res))
	{
		store_result(slot, festate->res);
		return slot;
	}

	/* We don't have any result even in remote server cursor. */
	ExecClearTuple(slot);
	return slot;
}

/*
 * oraReScanForeignScan
 *   - Restart this scan by resetting fetch location.
 */
static void
oraReScanForeignScan(ForeignScanState *node)
{
	OraFdwExecutionState *festate;

	festate = (OraFdwExecutionState *) node->fdw_state;

	if (festate->res);
		ORAclear(festate->res);

	execute_query(node);
}

/*
 * oraEndForeignScan
 *		Finish scanning foreign table and dispose objects used for this scan
 */
static void
oraEndForeignScan(ForeignScanState *node)
{
	OraFdwExecutionState *festate;

	festate = (OraFdwExecutionState *) node->fdw_state;

	/* if festate is NULL, we are in EXPLAIN; nothing to do */
	if (festate == NULL)
		return;

	if (festate->res);
		ORAclear(festate->res);

	//ReleaseConnection(festate->conn);
}

/*
 * Estimate costs of scanning a foreign table.
 */
static void
estimate_costs(PlannerInfo *root, RelOptInfo *baserel,
			   const char *sql,
			   Cost *startup_cost, Cost *total_cost)
{
	/* TODO Selectivity of quals pushed down should be considered. */

	*startup_cost += CONNECTION_COSTS;
	*total_cost += CONNECTION_COSTS;

	/* add cost to transfer result. */
	*total_cost += TRANSFER_COSTS_PER_BYTE * baserel->width * baserel->tuples;
	*total_cost += cpu_tuple_cost * baserel->tuples;
}

/*
 * Execute remote query with current parameters.
 */
static void
execute_query(ForeignScanState *node)
{
	OraFdwExecutionState *festate;
	ORAconn		   *conn;
	FdwPlan		   *fdwplan;
	char		   *sql;
	Oid				relid;
	ForeignTable   *table;
	const char	   *value;
	int				fetchsize;
	int				max_value_len;
	ORAresult	   *res;
	MemoryContext	oldcontext;

	festate = (OraFdwExecutionState *) node->fdw_state;

	/*
	 * Execute remote query with parameters.
	 */
	conn = festate->conn;
	fdwplan = ((ForeignScan *) node->ss.ps.plan)->fdwplan;
	sql = strVal(list_nth(fdwplan->fdw_private, 0));

	relid = RelationGetRelid(node->ss.ss_currentRelation);
	table = GetForeignTable(relid);

	value = GetFdwOptionList(table->options, "fetchsize");
	fetchsize = value ? parse_int_value(value, "fetchsize") :
							ORAfetchsize(conn);
	fetchsize = (fetchsize > 0) ? fetchsize : DEFAULT_FETCHSIZE;

	value = GetFdwOptionList(table->options, "max_value_len");
	max_value_len = value ? parse_int_value(value, "max_value_len") :
							ORAmax_value_len(conn);

	/* -1 is the terminal null character of the fetch buffer. */
	if (max_value_len > MAX_DATA_SIZE - 1)
		ereport(ERROR,
				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
				 errmsg("max_value_len cannot exceed %d", MAX_DATA_SIZE - 1)));

	oldcontext = MemoryContextSwitchTo(node->ss.ps.state->es_query_cxt);
	res = ORAopen(conn, sql, fetchsize, max_value_len);
	MemoryContextSwitchTo(oldcontext);

	/*
	 * If the query has failed, reporting details is enough here.
	 * Connections which are used by this query (including other scans) will
	 * be cleaned up by the foreign connection manager.
	 */
	if (!res || ORAresultStatus(res) != PGRES_TUPLES_OK)
	{
		char *msg;

		msg = pstrdup(ORAerrorMessage(conn));
		ORAclear(res);
		ereport(ERROR,
				(errmsg("could not execute foreign query"),
				 errdetail("%s", msg),
				 errhint("%s", sql)));
	}

	oldcontext = MemoryContextSwitchTo(node->ss.ps.state->es_query_cxt);
	TupleFormerInit(ORAgetTupleFormer(res),
		node->ss.ss_ScanTupleSlot->tts_tupleDescriptor);
	MemoryContextSwitchTo(oldcontext);

	festate->res = res;
}

/*
 * Create tuples from ORAresult and store them into slot.
 */
static void
store_result(TupleTableSlot *slot, ORAresult *res)
{
	int					i;
	int					nfields;
	Datum	   *dvalues;
	bool	   *nulls;
	Form_pg_attribute  *attrs;
	TupleDesc	tupdesc = slot->tts_tupleDescriptor;

	nfields = ORAnfields(res);
	attrs = tupdesc->attrs;

	/* buffer should include dropped columns */
	dvalues = (Datum *) palloc(tupdesc->natts * sizeof(Datum));
	nulls = (bool *) palloc(tupdesc->natts * sizeof(bool));

	/* put all tuples into the tuplestore */
	{
		int			j;
		HeapTuple	tuple;

		for (i = 0, j = 0; i < tupdesc->natts; i++)
		{
			/* skip dropped columns. */
			if (attrs[i]->attisdropped)
			{
				nulls[i] = true;
				continue;
			}

			if (ORAgetisnull(res, j))
				nulls[i] = true;
			else
			{
				char   *value;
				int		size;

				value = ORAgetvalue(res, j);
				size = ORAgetamountsize(res, j);
				dvalues[i] = TupleFormerValue(ORAgetTupleFormer(res),
								value, size, j);
				nulls[i] = false;
			}

			j++;
		}

		/*
		 * Build the tuple and put it into the slot.
		 * We don't have to free the tuple explicitly because it's been
		 * allocated in the per-tuple context.
		 */
		tuple = heap_form_tuple(tupdesc, dvalues, nulls);
		ExecStoreTuple(tuple, slot, InvalidBuffer, false);
	}

	pfree(dvalues);
	pfree(nulls);
}

void
ora_ereport(int sqlstate,
			 const char *message,
			 const char *detail,
			 const char *hint,
			 const char *context)
{
	ereport(ERROR, (errcode(sqlstate),
		message ? errmsg("%s", message) : errmsg("unknown error"),
		detail ? errdetail("%s", detail) : 0,
		hint ? errhint("%s", hint) : 0,
		context ? errcontext("%s", context) : 0));
}

void
ora_elog(int level, const char *message)
{
	elog(level, "%s", message);
}
