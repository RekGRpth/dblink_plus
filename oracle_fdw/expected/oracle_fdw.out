\c contrib_regression_utf8
SET DateStyle = 'Postgres, YMD';
-- プラン表示
EXPLAIN (COSTS FALSE) SELECT * FROM ft1 ORDER BY c1;
                                QUERY PLAN                                
--------------------------------------------------------------------------
 Sort
   Sort Key: c1
   ->  Foreign Scan on ft1
         Remote SQL: SELECT "c1", "c2", "c3", "c4" FROM "oracle_fdw"."t1"
(4 rows)

-- simple query
ALTER USER MAPPING FOR PUBLIC SERVER ora_server OPTIONS (SET user 'invalid_user');
SELECT * FROM ft1 ORDER BY c1; -- username ERROR
ERROR:  could not establish connection
DETAIL:  OCI_ERROR - ORA-01017: ユーザー名/パスワードが無効です。ログオンは拒否されました。

ALTER USER MAPPING FOR PUBLIC SERVER ora_server OPTIONS (DROP user);
SELECT * FROM ft1 ORDER BY c1;
ERROR:  user is required
ALTER USER MAPPING FOR PUBLIC SERVER ora_server OPTIONS (ADD user 'oracle_fdw');
SELECT * FROM ft1 ORDER BY c1;
 c1 |  c2   |  c3   |            c4            
----+-------+-------+--------------------------
  1 | A     | a     | Sat Jan 01 12:34:56 2011
  2 | BB    | bb    | Sun Jan 02 12:34:56 2011
  3 | CCC   | ccc   | Mon Jan 03 12:34:56 2011
  4 | DDDD  | dddd  | Tue Jan 04 12:34:56 2011
  5 | あ    | ア    | Wed Jan 05 12:34:56 2011
  6 |   A   | abcba | Wed Jan 05 12:34:56 2011
(6 rows)

ALTER SERVER ora_server OPTIONS (SET dbname 'aaa');
SELECT * FROM ft1 ORDER BY c1;
ERROR:  could not connect to server "ora_server"
DETAIL:  OCI_ERROR - ORA-12154: TNS: 指定された接続識別子を解決できませんでした

ALTER SERVER ora_server OPTIONS (DROP dbname);
SELECT * FROM ft1 ORDER BY c1;
 c1 |  c2   |  c3   |            c4            
----+-------+-------+--------------------------
  1 | A     | a     | Sat Jan 01 12:34:56 2011
  2 | BB    | bb    | Sun Jan 02 12:34:56 2011
  3 | CCC   | ccc   | Mon Jan 03 12:34:56 2011
  4 | DDDD  | dddd  | Tue Jan 04 12:34:56 2011
  5 | あ    | ア    | Wed Jan 05 12:34:56 2011
  6 |   A   | abcba | Wed Jan 05 12:34:56 2011
(6 rows)

ALTER FOREIGN TABLE ft1 OPTIONS (SET relname 't2');
EXPLAIN (COSTS FALSE) SELECT * FROM ft1 ORDER BY c1;
                                QUERY PLAN                                
--------------------------------------------------------------------------
 Sort
   Sort Key: c1
   ->  Foreign Scan on ft1
         Remote SQL: SELECT "c1", "c2", "c3", "c4" FROM "oracle_fdw"."t2"
(4 rows)

SELECT * FROM ft1 ORDER BY c1; -- colname ERROR
ERROR:  OCI_ERROR - ORA-00904: "c4": 無効な識別子です。

ALTER FOREIGN TABLE ft1 OPTIONS (SET relname 't1');
-- char(5) length
EXPLAIN (COSTS FALSE) SELECT c1, c2, length(c2), octet_length(c2), c3, length(c3), octet_length(c3), c4 FROM ft1 ORDER BY c1;
                                QUERY PLAN                                
--------------------------------------------------------------------------
 Sort
   Sort Key: c1
   ->  Foreign Scan on ft1
         Remote SQL: SELECT "c1", "c2", "c3", "c4" FROM "oracle_fdw"."t1"
(4 rows)

SELECT c1, c2, length(c2), octet_length(c2), c3, length(c3), octet_length(c3), c4 FROM ft1 ORDER BY c1;
 c1 |  c2   | length | octet_length |  c3   | length | octet_length |            c4            
----+-------+--------+--------------+-------+--------+--------------+--------------------------
  1 | A     |      1 |            1 | a     |      5 |            5 | Sat Jan 01 12:34:56 2011
  2 | BB    |      2 |            2 | bb    |      5 |            5 | Sun Jan 02 12:34:56 2011
  3 | CCC   |      3 |            3 | ccc   |      5 |            5 | Mon Jan 03 12:34:56 2011
  4 | DDDD  |      4 |            4 | dddd  |      5 |            5 | Tue Jan 04 12:34:56 2011
  5 | あ    |      1 |            3 | ア    |      3 |            5 | Wed Jan 05 12:34:56 2011
  6 |   A   |      5 |            5 | abcba |      5 |            5 | Wed Jan 05 12:34:56 2011
(6 rows)

-- query with projection
EXPLAIN (COSTS FALSE) SELECT c1 FROM ft1 ORDER BY c1;
                                QUERY PLAN                                
--------------------------------------------------------------------------
 Sort
   Sort Key: c1
   ->  Foreign Scan on ft1
         Remote SQL: SELECT "c1", NULL, NULL, NULL FROM "oracle_fdw"."t1"
(4 rows)

SELECT c1 FROM ft1 ORDER BY c1;
 c1 
----
  1
  2
  3
  4
  5
  6
(6 rows)

EXPLAIN (COSTS FALSE) SELECT c1, length(c2), length(c3) FROM ft2 ORDER BY c1;
                             QUERY PLAN                             
--------------------------------------------------------------------
 Sort
   Sort Key: c1
   ->  Foreign Scan on ft2
         Remote SQL: SELECT "c1", "c2", "c3" FROM "oracle_fdw"."t2"
(4 rows)

SELECT c1, length(c2), length(c3) FROM ft2 ORDER BY c1;
 c1 | length | length 
----+--------+--------
  1 |   2000 |   1000
  2 |   4000 |   3000
(2 rows)

-- join two foreign tables
EXPLAIN (COSTS FALSE) SELECT * FROM ft1 JOIN ft2 ON (ft1.c1 = ft2.c1) ORDER BY ft1.c1;
                                   QUERY PLAN                                   
--------------------------------------------------------------------------------
 Merge Join
   Merge Cond: (ft1.c1 = ft2.c1)
   ->  Sort
         Sort Key: ft1.c1
         ->  Foreign Scan on ft1
               Remote SQL: SELECT "c1", "c2", "c3", "c4" FROM "oracle_fdw"."t1"
   ->  Sort
         Sort Key: ft2.c1
         ->  Foreign Scan on ft2
               Remote SQL: SELECT "c1", "c2", "c3" FROM "oracle_fdw"."t2"
(10 rows)

--SELECT * FROM ft1 JOIN ft2 ON (ft1.c1 = ft2.c1) ORDER BY ft1.c1;
-- join itself
EXPLAIN (COSTS FALSE) SELECT * FROM ft1 t1 JOIN ft1 t2 ON (t1.c1 = t2.c1) ORDER BY t1.c1;
                                   QUERY PLAN                                   
--------------------------------------------------------------------------------
 Merge Join
   Merge Cond: (t1.c1 = t2.c1)
   ->  Sort
         Sort Key: t1.c1
         ->  Foreign Scan on ft1 t1
               Remote SQL: SELECT "c1", "c2", "c3", "c4" FROM "oracle_fdw"."t1"
   ->  Sort
         Sort Key: t2.c1
         ->  Foreign Scan on ft1 t2
               Remote SQL: SELECT "c1", "c2", "c3", "c4" FROM "oracle_fdw"."t1"
(10 rows)

SELECT * FROM ft1 t1 JOIN ft1 t2 ON (t1.c1 = t2.c1) ORDER BY t1.c1;
 c1 |  c2   |  c3   |            c4            | c1 |  c2   |  c3   |            c4            
----+-------+-------+--------------------------+----+-------+-------+--------------------------
  1 | A     | a     | Sat Jan 01 12:34:56 2011 |  1 | A     | a     | Sat Jan 01 12:34:56 2011
  2 | BB    | bb    | Sun Jan 02 12:34:56 2011 |  2 | BB    | bb    | Sun Jan 02 12:34:56 2011
  3 | CCC   | ccc   | Mon Jan 03 12:34:56 2011 |  3 | CCC   | ccc   | Mon Jan 03 12:34:56 2011
  4 | DDDD  | dddd  | Tue Jan 04 12:34:56 2011 |  4 | DDDD  | dddd  | Tue Jan 04 12:34:56 2011
  5 | あ    | ア    | Wed Jan 05 12:34:56 2011 |  5 | あ    | ア    | Wed Jan 05 12:34:56 2011
  6 |   A   | abcba | Wed Jan 05 12:34:56 2011 |  6 |   A   | abcba | Wed Jan 05 12:34:56 2011
(6 rows)

-- outer join
EXPLAIN (COSTS FALSE) SELECT * FROM ft1 t1 LEFT JOIN ft2 t2 ON (t1.c1 = t2.c1) ORDER BY 1,2,3,4,5,6;
                                      QUERY PLAN                                      
--------------------------------------------------------------------------------------
 Sort
   Sort Key: t1.c1, t1.c2, t1.c3, t1.c4, t2.c1, t2.c2
   ->  Merge Left Join
         Merge Cond: (t1.c1 = t2.c1)
         ->  Sort
               Sort Key: t1.c1
               ->  Foreign Scan on ft1 t1
                     Remote SQL: SELECT "c1", "c2", "c3", "c4" FROM "oracle_fdw"."t1"
         ->  Sort
               Sort Key: t2.c1
               ->  Foreign Scan on ft2 t2
                     Remote SQL: SELECT "c1", "c2", "c3" FROM "oracle_fdw"."t2"
(12 rows)

--SELECT * FROM ft1 t1 LEFT JOIN ft2 t2 ON (t1.c1 = t2.c1) ORDER BY 1,2,3,4,5,6;
-- WHERE clause
EXPLAIN (COSTS FALSE) SELECT * FROM ft1 WHERE c1 = 1 AND c2 < lower('AAA') AND c4 < now() and c4 < clock_timestamp();
                                                QUERY PLAN                                                
----------------------------------------------------------------------------------------------------------
 Foreign Scan on ft1
   Filter: ((c4 < now()) AND (c4 < clock_timestamp()))
   Remote SQL: SELECT "c1", "c2", "c3", "c4" FROM "oracle_fdw"."t1" WHERE (("c2" < 'aaa') AND ("c1" = 1))
(3 rows)

SELECT * FROM ft1 WHERE c1 = 1 AND c2 < lower('AAA') AND c4 < now() and c4 < clock_timestamp();
 c1 | c2 |  c3   |            c4            
----+----+-------+--------------------------
  1 | A  | a     | Sat Jan 01 12:34:56 2011
(1 row)

-- prepared statement
PREPARE st(int) AS SELECT * FROM ft1 WHERE c1 > $1 ORDER BY c1;
EXPLAIN (COSTS FALSE) EXECUTE st(1);
                                QUERY PLAN                                
--------------------------------------------------------------------------
 Sort
   Sort Key: c1
   ->  Foreign Scan on ft1
         Filter: (c1 > $1)
         Remote SQL: SELECT "c1", "c2", "c3", "c4" FROM "oracle_fdw"."t1"
(5 rows)

EXECUTE st(1);
 c1 |  c2   |  c3   |            c4            
----+-------+-------+--------------------------
  2 | BB    | bb    | Sun Jan 02 12:34:56 2011
  3 | CCC   | ccc   | Mon Jan 03 12:34:56 2011
  4 | DDDD  | dddd  | Tue Jan 04 12:34:56 2011
  5 | あ    | ア    | Wed Jan 05 12:34:56 2011
  6 |   A   | abcba | Wed Jan 05 12:34:56 2011
(5 rows)

EXPLAIN (COSTS FALSE) EXECUTE st(2);
                                QUERY PLAN                                
--------------------------------------------------------------------------
 Sort
   Sort Key: c1
   ->  Foreign Scan on ft1
         Filter: (c1 > $1)
         Remote SQL: SELECT "c1", "c2", "c3", "c4" FROM "oracle_fdw"."t1"
(5 rows)

EXECUTE st(2);
 c1 |  c2   |  c3   |            c4            
----+-------+-------+--------------------------
  3 | CCC   | ccc   | Mon Jan 03 12:34:56 2011
  4 | DDDD  | dddd  | Tue Jan 04 12:34:56 2011
  5 | あ    | ア    | Wed Jan 05 12:34:56 2011
  6 |   A   | abcba | Wed Jan 05 12:34:56 2011
(4 rows)

DEALLOCATE st;
-- change datastyle
SET DATESTYLE = 'Postgres, MDY';
SELECT * FROM ft1 ORDER BY c4;
 c1 |  c2   |  c3   |            c4            
----+-------+-------+--------------------------
  1 | A     | a     | Sat Jan 01 12:34:56 2011
  2 | BB    | bb    | Sun Jan 02 12:34:56 2011
  3 | CCC   | ccc   | Mon Jan 03 12:34:56 2011
  4 | DDDD  | dddd  | Tue Jan 04 12:34:56 2011
  5 | あ    | ア    | Wed Jan 05 12:34:56 2011
  6 |   A   | abcba | Wed Jan 05 12:34:56 2011
(6 rows)

SET DATESTYLE = 'Postgres, DMY';
SELECT * FROM ft1 ORDER BY c4;
 c1 |  c2   |  c3   |            c4            
----+-------+-------+--------------------------
  1 | A     | a     | Sat 01 Jan 12:34:56 2011
  2 | BB    | bb    | Sun 02 Jan 12:34:56 2011
  3 | CCC   | ccc   | Mon 03 Jan 12:34:56 2011
  4 | DDDD  | dddd  | Tue 04 Jan 12:34:56 2011
  5 | あ    | ア    | Wed 05 Jan 12:34:56 2011
  6 |   A   | abcba | Wed 05 Jan 12:34:56 2011
(6 rows)

-- timestamp to text
ALTER FOREIGN TABLE ft1 ALTER c4 TYPE text;
-- simple query
EXPLAIN (COSTS FALSE) SELECT * FROM ft1 ORDER BY c1;
                                QUERY PLAN                                
--------------------------------------------------------------------------
 Sort
   Sort Key: c1
   ->  Foreign Scan on ft1
         Remote SQL: SELECT "c1", "c2", "c3", "c4" FROM "oracle_fdw"."t1"
(4 rows)

SELECT * FROM ft1 ORDER BY c1;
 c1 |  c2   |  c3   |              c4               
----+-------+-------+-------------------------------
  1 | A     | a     | 2011-01-01 AD 12:34:56.000000
  2 | BB    | bb    | 2011-01-02 AD 12:34:56.000000
  3 | CCC   | ccc   | 2011-01-03 AD 12:34:56.000000
  4 | DDDD  | dddd  | 2011-01-04 AD 12:34:56.000000
  5 | あ    | ア    | 2011-01-05 AD 12:34:56.000000
  6 |   A   | abcba | 2011-01-05 AD 12:34:56.000000
(6 rows)

-- max_valu_len
ALTER SERVER ora_server OPTIONS (ADD max_value_len '1073741823');
ERROR:  invalid value for max_value_len: "1073741823"
ALTER SERVER ora_server OPTIONS (ADD max_value_len '1073741822');
ALTER SERVER ora_server OPTIONS (DROP max_value_len);
